### ISSUE TITLE
SyntaxError: Unexpected end of JSON input

### ISSUE THREAD

**Issue by bbuhler** [pocketjoso/penthouse#148]
Hi,

can you please help me to figure out what the problem is. 
It appears super randomly. But leads to break every 2nd/3rd build.

```
[09:23:20] Starting 'critical'...
caught err SyntaxError: Unexpected end of JSON input
    at Object.parse (native)
    at postformatting (/home/bbuhler/Dokumente/liita-client/node_modules/penthouse/lib/postformatting/index.js:25:28)
    at /home/bbuhler/Dokumente/liita-client/node_modules/penthouse/lib/index.js:129:26
    at ChildProcess.<anonymous> (/home/bbuhler/Dokumente/liita-client/node_modules/penthouse/lib/index.js:128:9)
    at emitTwo (events.js:106:13)
    at ChildProcess.emit (events.js:191:7)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:215:12)
Fatal undefined

Process finished with exit code 2
```

**Comment by pocketjoso** [comment]
Hi,
any change you can share the css and/or the url you're running this for?

**Comment by bbuhler** [comment]
I can not share the hole project with you.

I tried creating a test project with just the HTML and CSS files and the critical task, but in that project it isn't reproducible (see [critical-test.zip](https://github.com/pocketjoso/penthouse/files/661188/critical-test.zip)).

It feels like the error happens when the critical task runs in parallel with some other tasks like imagemin.

**Comment by pocketjoso** [comment]
Fair enough. I'll look to see whether there's anything suspicious in my code that can explain your problem.

**Comment by pocketjoso** [comment]
It's hard for me to see what's going on here without a re-producable test case.

Could you add a `try/catch` around [the relevant line](https://github.com/pocketjoso/penthouse/blob/master/lib/postformatting/index.js#L25) of your _local_ penthouse node_module on your file system, and log the value of `cssAstRulesJsonString` when it crashes? Sounds like it's not a JSON, but would be interesting to see what it is then... If you could share that output, if not in exact details, at least in words.

**Comment by bbuhler** [comment]
```
[14:16:30] Starting 'critical'...
debuggingHelp:  
time: 2393 | postformatting: remove phantom js security errors
stdErr:  Unable to open config: "/home/bbuhler/Dokumente/liita-client/node_modules/penthouse/lib/phantomjs/config.json"
time: 2393 | ERR: postformatting: SyntaxError: Unexpected end of JSON input

stdOut:  
stdOutString:  
cssAstRulesJsonString:  
```

`cssAstRulesJsonString` seems to be like an empty string.
But note the `stdErr`. What config.json is missing there and how can that happen?

**Comment by pocketjoso** [comment]
Yeah the missing `package.json` should be unrelated (and is also fixed by the latest version of penthouse; if you re-install/upgrade `critical`).

How did you look this? Did you use that internal `debuglog` function? And where did you get the `stdOut` from? from the call site? (where postformatting was called from)?

**Comment by bbuhler** [comment]
> How did you look this? Did you use that internal `debuglog` function? 

I tried `console.log` as well as the internal `debuglog`.

> And where did you get the `stdOut` from? from the call site? (where postformatting was called from)?

Yes, from inside the `cp.on('exit', ...` [here](https://github.com/pocketjoso/penthouse/blob/master/lib/index.js#L130)

So it seems like PhantomJS is not returning any output.

**Comment by pocketjoso** [comment]
can you do a `typeof stdOut`? is it `String` or `undefined`?

**Comment by pocketjoso** [comment]
Next, can you log what (full original) css Penthouse is running with [here](https://github.com/pocketjoso/penthouse/blob/master/lib/index.js#L205). You don't have to share it with me, but is it a String, and does it contain your css?

**Comment by pocketjoso** [comment]
At this point, of course I can add a guard so your program won't crash, but it would just return an empty critical css file if so, which I presume is also an incorrect answer here. So let's figure out why the `stdOut` is empty.

**Comment by bbuhler** [comment]
> can you do a typeof stdOut? is it String or undefined?
> 
> Edit: well, it should really be a String.

Yes, it is.

> Next, can you log what (full original) css Penthouse is running with here. You don't have to share it with me, but is it a String, and does it contain your css?

It is a String too and contains the CSS.

**Comment by bbuhler** [comment]
I enabled the `m.DEBUG` in `lib/index.js` too which gave's me some more output:
```
time: 821 | parsed ast (without errors)
time: 1426 | Penthouse core startopened ast from fileparsed ast from jsongetCriticalPathCssprepareNewPagestripped out non matching media queries
time: 1697 | page openedpruneNonCriticalCss
time: 1953 | processCssRules BEFOREprocessCssRules AFTERGENERATION_DONEreturnCssFromAstRulesfinalCss: write - DONE!
```

**Comment by pocketjoso** [comment]
@bbuhler This is the relevant code (from [here](https://github.com/pocketjoso/penthouse/blob/master/lib/phantomjs/core.js#L91-L93):
```js
stdout.write(JSON.stringify(criticalRules))
debuglog('finalCss: write - DONE!')
phantomExit(0)
```
We know that line 2 and 3 happens for you, but it seems like there is some issue with the PhantomJS' `stdout.write`, causing your bug. We know that you have `criticalRules`, otherwise this code branch wouldn't execute, but to check for yourself, you change the debuglog to also log the `criticalRules` locally.

I wonder if for some reason in your case the `stdout.write` call is not _not_ blocking, hence not done before the `phantomExit` call in your case. I don't know any elegant way of forcing it to be blocked in PhantomJS  code, perhaps you could test just wrapping the `phantomExit` call here in a `setTimeout` - just to see if any `delay` value (even 0) makes a difference for your bug.

Finally, what `node` version are you on?

**Comment by pocketjoso** [comment]
You could also try inserting `stdout.write('before|after')` lines surrounding the real stdout.write line here, to see if those come through (if you log the stdOut on the other end).

It could also be beneficial to `debuglog` `JSON.stringify(criticalRules)`, to verify that it is indeed a JSON, in your case.

---

Obviously for all of this, it is during the runs when your process crashes that we're interested in what values we see along the way.

Let me know what you find.

**Comment by noderat** [comment]
I'm getting the exact same error using critical. It appears the problem happens on every build when I pass a large array of resolutions from critical.

```
caught err SyntaxError: Unexpected end of JSON input
    at Object.parse (native)
    at postformatting (/data/jenkins/jobs/--redacted---/workspace/node_modules/penthouse/lib/postformatting/index.js:25:28)
    at /data/jenkins/jobs/---redacted---/workspace/node_modules/penthouse/lib/index.js:133:26
    at ChildProcess.<anonymous> (/data/jenkins/jobs/---redacted---/workspace/node_modules/penthouse/lib/index.js:132:9)
    at emitTwo (events.js:106:13)
    at ChildProcess.emit (events.js:191:7)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:215:12)
```

Here's what I've been testing:
```javascript
    dimensions: [
      // { width: 1280, height: 1024 },
      // { width: 1200, height: 1024 },
      // { width: 1050, height: 768 },
      { width: 768, height: 375 },
      { width: 650, height: 375 },
      { width: 600, height: 375 },
      { width: 500, height: 375 },
      { width: 400, height: 375 },
    ],
```

As it stands that example dimensions array causes it to fail sporadically, if I remove the comments it fails on every build. Removing the dimensions causes all attempts to pass.

**Comment by pocketjoso** [comment]
@noderat regarding debugging this I can only repeat what I said before.
1. First make sure you run `penthouse` in debug mode. I've never tested this via critical, and there's no direct support there, but it might work to just add this code to where you call `critical` from:
```js
var penthouse = require('penthouse')
penthouse.DEBUG = true
```
2.  follow this advise: https://github.com/pocketjoso/penthouse/issues/148#issuecomment-269083461
(i.e. locate your local `penthouse` folder in your `node_modules`, and go into the source and modify it according to that comment.
---
I can't help you debug any further without reduced test cases, including css + html/url.

**Comment by pocketjoso** [comment]
@noderat Another question however - why are you generating critical css for dimensions like that? are you aware that as long as your css is responsive (using media queries), your largest dimension critical css will contain all the styles for all the smaller dimensions, i.e. that the smaller ones will just be subsets of the larger ones? There's no need to separate by dimensions, other than if you were really trying to shave of some extra large screen styles from a small screen critical css.

**Comment by bbuhler** [comment]
>  debuglog to also log the criticalRules (on your local installation of Penthouse, on your file system).

It contains the AST object.

> Finally, what node version are you on?

Local v6.9.3, on Jenkins it's v4.2.x. The bug appears with both versions.

> It could also be beneficial to debuglog JSON.stringify(criticalRules), to verify that it is indeed a JSON, in your case.

It is a JSON.

> I wonder if for some reason in your case the stdout.write call is not not blocking, hence not done before the phantomExit. I don't know any elegant way of forcing it to be blocked, or waiting for it to be finished, in PhantomJS code, perhaps you could test just wrapping the phantomExit call here in a setTimeout - just to see if any delay value (even 0) makes your bug disappear.

It seems to be something in that direction.

As you can see in that [gulp-critical-error-06-01-17.log](https://github.com/pocketjoso/penthouse/files/689072/gulp-critical-error-06-01-17.log.txt) the error appears even before the `stdout.write` or `phantomExit` gets called. Or can that caused by some output delay? (Because some log messages a concatenated)

A timeout delay of 0 didn't help.

> You could also try inserting stdout.write('before|after') lines surrounding the real stdout.write line here, to see if those come through (if you log the stdOut on the other end).

I will try that later too.

----

Btw. I don't use the `dimensions` option, like @noderat.

**Comment by pocketjoso** [comment]
Just an update that I can now reproduce this, and am looking into it.

**Comment by noderat** [comment]
Sorry, winter storm kept me out of the office. My snippet was an example of a way I could reproduce the issue not actual production code.

**Comment by pocketjoso** [comment]
Turns out I could only reproduce it very inconsistently. Seems to be some kind of interference between parallel system.out.write calls, but I haven't found any more information about this, yet. I'll keep looking.

If anyone manages to find a reduced reproducable test case with a public url/html file and css, that would make this a lot easier to really understand.

**Comment by noderat** [comment]
This is also the problem I'm experiencing as in my example I could only increase the likelihood of the failure happening by adding additional dimensions to critical's configuration but it still could not get it to happen on every build.

**Comment by arminnaimi** [comment]
I experience the same issue, but only in my 14.04 LTS 64-bit Ubuntu GitLab CI runner. This error does not occur on my Windows 7 or macOS Sierra machine. I hope this info may be helpful.

**Comment by forrestmid** [comment]
Thought I'd put it out there that I have been able to reproduce this issue in an inconsistently consistent manner utilizing an AWS Lambda function I built. I couldn't duplicate it as consistently as would be nice to have, but on pcworld.com I got it about 2% of the time.

Basically what I needed to do was trigger about 50 requests utilizing Penthouse on multiple Lambda invocations.  An invocation errors out at least once or twice almost every time. The key, I believe, is to invoke multiple Penthouse calls concurrently within each function. I'm using the `async` library and that was throwing up an error often enough.

I've still not been able to duplicate the issue in my dev environment, but I at least know that pcworld.com is throwing an error when running it through my Lambda function.

**Comment by pocketjoso** [comment]
@forrestmid thanks for adding to the knowledge pool. My situation is still the same - I don't know what causes the issue, or how to fix it, unfortunately.

**Comment by forrestmid** [comment]
Well, I can provide the Lambda function if you'd like to mess around with it on there. Unfortunately the native phantomjs code has to be precompiled, so I'm not sure if there would be problems in recompiling it while you make changes and test. It's a slow process though. With phantom it's almost 40 megs zipped, so it's a change, upload, test, change, upload, test, kind of slowly. 

The only thing that I can think of it being associated with is the events and, somehow, you're putting out global events which are being caught by the other Penthouse process. The reason, I'd think, that it only happens some of the time is because that event listener on the other Penthouse process only happens to be associated at the same time as the original, thus throwing an unexpected end because the `criticalRules = JSON.parse(cssAstRulesJsonString)` isn't ready to be called yet on the second process. If it's sending the events out globally, as compared to only inside of some scope that Penthouse processes run in, then if some sort of `event.on('foo')` was called simultaneously, or close to simultaneously, on two or more Penthouse functions, then when one finishes they all think they're finished and call a `JSON.parse` on an output that hasn't yet finished constructing.

I could also totally be off base and talking out my butt, but I have figured out a workaround which is to just not run the Penthouse processes concurrently in my code, instead running through them in series. Not a huge deal, but it's kind of annoying and definitely increases my latency on each function by several seconds.
 
It must, therefore, be associated with parallel executions with Penthouse and, due to the stack trace containing an event call, I'd assume it has to do with two event listeners getting called in separate functions simultaneously.

Just my two cents. :D

**Comment by hummal** [comment]
Just for your interest: Had the same error. Commented out the phantomJsOptions and it worked.
Maybe there is something with translating these options into json format.

**Comment by pocketjoso** [comment]
Philipp,
Could you please verify that "fix" is reproducable - run penthouse at least
5 time each with/without the fix and see if it always works with and breaks
without. Of so this is unexpected, but good news.

Next step of so would be if you could disable one option at a time and find
which specific option "fixes" it by commenting.


On 16 Mar 2017 16:44, "Philipp Winterle" <notifications@github.com> wrote:

Just for your interest: Had the same error. Commented out the
phantomJsOptions and it worked.
Maybe there is something with translating these options into json format.

—
You are receiving this because you commented.

Reply to this email directly, view it on GitHub
<https://github.com/pocketjoso/penthouse/issues/148#issuecomment-287099048>,
or mute the thread

**Comment by hummal** [comment]
Could reproduce it every time. So yeah

**Comment by pocketjoso** [comment]
@hummal
the correct syntax here would be `"output-encoding": "utf8"`. But that's also the default value, not sure if you are aware?

**Comment by hummal** [comment]
Thanks for the update. 
With the correct property key and value it doesn't crash anymore.

**Comment by michalkleiner** [comment]
I'm seeing the same error in my virtualised environment. It doesn't happen every time and I haven't been able to reproduce it reliably.
When the error occurs, `stdOutString` and `cssAstRulesJsonString` in `postformatting/index.js` are both empty strings.

**Comment by pocketjoso** [comment]
@michalkleiner It happens when you have multiple penthouse (more specifically: phantomjs) processes running at the same time. I don't have a solution for it, but you can always avoid the problem on your end by only running one `penthouse` job at a time.

**Comment by michalkleiner** [comment]
Hi @pocketjoso - yes, that is what I came yesterday to as well - we had multiple critical CSS generating gulp tasks running in parallel. When I changed the setup to use https://www.npmjs.com/package/run-sequence and run them one after another, the issue didn't appear.

**Comment by hummal** [comment]
@michalkleiner You should use nodejs workers to use penthouse in parallel. Works like charm

**Comment by pocketjoso** [comment]
@hummal are you sure? each `penthouse` process already uses child_processes; it shouldn't make any difference. If it does however then I need to explore why, so the fix can be brought into Penthouse,

**Comment by hummal** [comment]
@pocketjoso Iam using it that way. Calculating the amount of spawned workers with the amount of cpus and go. Never had that error again after fixing the correct syntax. And on the biggest machine I have 32 workers. So yeah

**Comment by phaistonian** [comment]
@hummal Can you provide an example of how this can be done?

**Comment by pocketjoso** [comment]
This was a phantomJS specific issue; it is gone in #174, which uses Chrome headless. Will close when that PR is merged, should happen late this week.

**Comment by pocketjoso** [comment]
`Penthouse 1.0` landed (via #174), please try upgrading. That should resolve this issue, if it doesn't please comment again and I'll re-open.
